includes/class-boldgrid-inspirations-onboarding-progress.php

```diff
@@ -0,0 +1,209 @@
+
+	/**
+	 * The onboarding progress option name.
+	 *
+	 * @since 2.8.0
+	 *
+	 * @var string
+	 */
+	var $progress_option_name;
+
+	/**
+	 * Inspirations Configs
+	 *
+	 * @since 2.8.0
+	 *
+	 * @var array
+	 */
+	var $configs;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 2.8.0
+	 */
+	public function __construct( $configs ) {
+		$this->configs              = $configs;
+		$this->tasks_option_name    = $this->configs['onboarding_tasks_option'];
+		$this->progress_option_name = $this->configs['onboarding_progress_option'];
+	}
+
+	/**
+	 * Add Admin Ajax hooks.
+	 *
+	 * @since 2.8.0
+	 */
+	public function add_ajax_hooks() {
+		add_action( 'wp_ajax_boldgrid_inspirations_update_task', array( $this, 'ajax_update_task' ) );
+	}
+
+	/**
+	 * Get Task.
+	 *
+	 * @since 2.8.0
+	 *
+	 * @param string $task_id The task ID.
+	 *
+	 * @return array The task object.
+	 */
+	public function get_task( $task_id ) {
+		$tasks = get_option( $this->tasks_option_name );
+
+		if ( ! is_array( $tasks ) ) {
+			return false;
+		}
+
+		foreach ( $tasks as $task ) {
+			if ( $task['id'] === $task_id ) {
+				return $task;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Update Task.
+	 *
+	 * @since 2.8.0
+	 *
+	 * @param array $task The task object.
+	 */
+	public function update_task( $task ) {
+		$tasks = get_option( $this->tasks_option_name );
+
+		if ( ! is_array( $tasks ) ) {
+			return;
+		}
+
+		foreach ( $tasks as $key => $task_data ) {
+			if ( $task_data['id'] === $task['id'] ) {
+				$tasks[ $key ] = $task;
+			}
+		}
+
+		update_option( $this->tasks_option_name, $tasks );
+
+		$this->update_percent_complete();
+	}
+
+	/**
+	 * Update Percent Complete.
+	 *
+	 * @since 2.8.0
+	 */
+	public function update_percent_complete() {
+		$tasks = get_option( $this->tasks_option_name );
+
+		if ( ! is_array( $tasks ) ) {
+			return;
+		}
+
+		$complete = 0;
+		$total    = count( $tasks );
+
+		foreach ( $tasks as $task ) {
+			if ( $task['task_complete'] ) {
+				$complete++;
+			}
+		}
+
+		$percent_complete = ( $total > 0 ) ? $complete / $total : 0;
+
+		if ( get_option( $this->progress_option_name ) !== $percent_complete ) {
+			do_action( 'boldgrid_feedback_add', 'onboarding_task_completed', $percent_complete, true );
+			update_option( $this->progress_option_name, $percent_complete );
+		}
+	}
+
+	/**
+	 * Ajax callback to update the status of a task.
+	 *
+	 * @since 2.8.0
+	 */
+	public function ajax_update_task() {
+		if ( ! isset( $_POST['nonce'] ) ) {
+			wp_send_json_error( 'missing_nonce' );
+		}
+
+		if ( ! wp_verify_nonce( sanitize_text_field( wp_unslash( $_POST['nonce'] ) ), 'boldgrid_inspirations_update_task' ) ) {
+			wp_send_json_error( 'invalid_nonce' );
+		}
+
+		if ( ! current_user_can( 'manage_options' ) ) {
+			wp_send_json_error( 'invalid_permissions' );
+		}
+
+		$task_id = isset( $_POST['task_id'] ) ? sanitize_text_field( wp_unslash( $_POST['task_id'] ) ) : '';
+
+		if ( 'skip_all_tasks' === $task_id ) {
+			$this->skip_all_tasks();
```

Comment:
If someone is skipping all tasks, can we return at this point?

Reply:
OK, I see the `skip_all_tasks()` function has `wp_send_json_success();`. I would recommend not adding `wp_send_json_success();` to a function that is not solely for handling an ajax request.

Reply:
How would you recommend this then? 
This whole thing is about handling the ajax request. If the $task_id is 'skip_all_tasks' then we want it to execute the code in the skip_all_tasks() method, and then send the success, thus ending the ajax call, and short circuiting the rest of this method